lib/commons/aria/roles.js

```diff
@@ -204,3 +202,252 @@ aria.implicitRole = function(node) {
+		}
+		if (!el.attributes) {
+			return;
+		}
+		el.attributes.forEach(attr => {
+			if (!nodeAttrs.includes(attr.name)) {
+				return;
+			}
+			if (typeof attr.value === 'undefined') {
+				isApplicable = true;
+				return;
+			}
+			if (tagName === 'IMG' && attr.value === '') {
+				isApplicable = true;
+				return;
+			}
+			if (!node.getAttribute(attr.name)) {
+				return;
+			}
+			const nodeAttrValue = node
+				.getAttribute(attr.name)
+				.trim()
+				.toUpperCase();
+			if (attr.value === nodeAttrValue) {
+				isApplicable = true;
+				return;
+			}
+		});
+	});
+	return isApplicable;
+}
+
+/**
+ * validate if a given role is an allowed ARIA role.
+ * @method validateAllowedAriaRole
+ * @param {String} tagName tag name of a node
+ * @param {String} role aria role to check
+ * @return {Boolean} retruns true/false
+ * @private
+ */
+function validateAllowedAriaRole(node, tagName, role) {
+	/**
+	 * Collection of functions, which evaluate if a given ROLE is allowed on a specified type of ELEMENT.
+	 * @type {Object}
+	 */
+	const elementCheck = {
+		A: ({ node, isApplicable }) => {
+			if (node.namespaceURI === 'http://www.w3.org/2000/svg') {
+				return true;
+			}
+			if (node.href.length) {
+				return isApplicable;
+			}
+			return true;
+		},
+		AREA: ({ node }) => !node.href,
+		BUTTON: ({ node, role, isApplicable }) => {
+			if (node.getAttribute('type') === 'menu') {
+				return role === 'menuitem';
+			}
+			return isApplicable;
+		},
+		IMG: ({ node, isApplicable }) => {
+			if (node.alt) {
+				return !isApplicable;
+			}
+			return isApplicable;
+		},
+		INPUT: ({ node, role, isApplicable }) => {
+			switch (node.type) {
+				case 'button':
+				case 'image':
+					return isApplicable;
+				case 'checkbox':
+					if (role === 'button' && node.hasAttribute('aria-pressed')) {
+						return true;
+					}
+					return isApplicable;
+				case 'radio':
+					return role === 'menuitemradio';
+				default:
+					return false;
+			}
+		},
+		LI: ({ node, isApplicable }) => {
+			const hasImplicitListitemRole = node.matches('ol li, ul li');
+			if (hasImplicitListitemRole) {
+				return isApplicable;
+			}
+			return true;
+		},
+		LINK: ({ node }) => !node.href,
+		MENU: ({ node }) => {
+			if (node.getAttribute('type') === 'context') {
+				return false;
+			}
+			return true;
+		},
+		OPTION: ({ node }) => {
+			const withinOptionList = node.matches(
+				'select > option, datalist > option, optgroup > option'
+			);
+			return !withinOptionList;
+		},
+		SELECT: ({ node, role }) =>
+			!node.multiple && node.size <= 1 && role === 'menu',
+		SVG: ({ node, isApplicable }) => {
+			// if in svg context it all roles may be used
+			if (
+				node.parentNode &&
+				node.parentNode.namespaceURI === 'http://www.w3.org/2000/svg'
+			) {
+				return true;
+			}
+			return isApplicable;
+		}
+	};
+
+	const lookupTableRole = axe.commons.aria.lookupTable.role;
+	const canHaveNoRole = lookupTableRole.NO_ROLE.allowedElements.includes(
```

Comment:
I don't know where my previous comment went about not extending role with non-enumberable properties... but here it is again. I don't think this is a good idea. This quite unintuitive. We know there are users that manipulate the lookup table. This can lead to unpredictable and difficult to debug issues. Please put the `lookupTable.role.NO_ROLE.alloweDlement` array on `lookupTable.elmsAllowedNoRole`, and a similar thing for `elmsAllowedAnyRole`.